packages/wdio-browserstack-service/src/insights-handler.ts

```diff
@@ -323,14 +323,38 @@ class _InsightsHandler {
      */
 
     private attachHookData (context: any, hookId: string): void {
-        if (!context.currentTest || !context.currentTest.parent) {
+        if (context.currentTest && context.currentTest.parent) {
+            const parentTest = `${context.currentTest.parent.title} - ${context.currentTest.title}`
+            if (!this._hooks[parentTest]) {
+                this._hooks[parentTest] = []
+            }
+
+            this._hooks[parentTest].push(hookId)
             return
+        } else if (context.test) {
+            const setHooksFromSuite = (parent: any): boolean => {
+                if (!parent) {
+                    return false
+                }
+                for (const test of parent.tests) {
+                    const uniqueIdentifier = getUniqueIdentifier(test, this._framework)
+                    if (!this._hooks[uniqueIdentifier]) {
+                        this._hooks[uniqueIdentifier] = []
+                    }
+                    this._hooks[uniqueIdentifier].push(hookId)
+                    return true
+                }
+
+                for (const suite of parent.suites) {
+                    const result = setHooksFromSuite(suite)
+                    if (result) {
+                        return true
+                    }
+                }
+                return false
+            }
+            setHooksFromSuite(context.test.parent)
         }
```

Comment:
No need to create a `setHooksFromSuite` function if it is never be called elsewhere. Also, let's return early.
```suggestion
        if (context.currentTest && context.currentTest.parent) {
            const parentTest = `${context.currentTest.parent.title} - ${context.currentTest.title}`
            if (!this._hooks[parentTest]) {
                this._hooks[parentTest] = []
            }

            this._hooks[parentTest].push(hookId)
            return
        }
        
        if (!context.test || !context.test.parent) {
            return false
        }

        for (const test of context.test.parent.tests) {
            const uniqueIdentifier = getUniqueIdentifier(test, this._framework)
            if (!this._hooks[uniqueIdentifier]) {
                this._hooks[uniqueIdentifier] = []
            }
            this._hooks[uniqueIdentifier].push(hookId)
            return true
        }

        for (const suite of context.test.parent.suites) {
            const result = setHooksFromSuite(suite)
            if (result) {
                return true
            }
        }
        return false
```

Reply:
@christian-bromann it's a recursive function, that it is getting called from itself. It sets hooks recursively, so that's why had to make it a function

Reply:
Let's move it into a `utility.ts` file as it gets very hard to test it this way

Reply:
Have moved to a different function in that class, as that method requires the instance variables of the class. https://github.com/webdriverio/webdriverio/pull/10601/commits/5cfd96f8c5cb186ea62e97e0659b71e83c630b43